### PR TITLE
Bids 2857/notifications general tab

### PR DESCRIPTION
**Pull request by D13ce**
This PR
* Adds the components used to show the "General" tab in the "Manage Notifications" modal
* Adds the components used for the "Paired Devices" modal
* Adds logic to handle data coming from `/users/me/notifications/settings/general` (returning data for more than the "General" tab only), see `frontend/composables/notifications/useNotificationsManagementSettingsProvider.ts`. This can function as a blueprint for other parts this endpoint returns data for.
* Adds data types for everything related to the points above via `frontend/types/notifications/settings.ts`.
:warning: Please note that this has to be removed once the types are defined by the backend.
* Adds styling for a secondary button

This PR does not
* Implement the logic to handle user input in `frontend/components/notifications/management/NotificationsManagementPairedDevice.vue`. The reason for this is that when this PR was created, it was not completely sure what settings paired devices will even have.
* Show the "Linked App" line in the "Paired Devices" modal as discussed with both the designer and the PO

### CODE REVIEW

**Comment by marcel-bitfly** [comment]
The layout is jumping between these two states, would be nice if you count this in
image
Screenshot from 2024-07-24 13-16-28

File: frontend/i18n/en.json [hunk 1]
```diff
@@ -314,6 +314,31 @@
...
+    "general": {
+      "do_not_disturb": "Do not disturb",
+      "mute": {
+        "mutes_all": "Mutes all notifications",
+        "unmute": "Unmute",
+        "muted_until_date": "Muted until {date}",
+        "muted_until_turned_on": "Muted until turned back on",
+        "select_duration": "Select mute duration",
+        "count_hours": "{count} hour|{count} hours",
+        "until_turned_on": "Until I turn it back on"
```
Comment (marcel-bitfly) [inline frontend/i18n/en.json:326]: nitpick: (Imo a little more consistent, but naming things is hard)
```suggestion
      "mute": {
        "all": "Mutes all notifications",
        "unmute": "Unmute",
        "until": "Muted until {date}",
        "until_turned_on": "Muted until turned back on",
        "select_duration": "Select mute duration",
        "hours": "{count} hour|{count} hours",
        "until_turned_on": "Until I turn it back on"
```

File: frontend/components/notifications/management/NotificationsManagementGeneralTab.vue [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+      return $t('notifications.general.mute.muted_until_turned_on')
+    }
+    return $t('notifications.general.mute.muted_until_date', { date: formatTsToAbsolute(muteTimestamp.value, $t('locales.date'), true) })
+  }
+})
+
+</script>
+
+<template>
+  <NotificationsManagementPairedDevicesModal v-model="pairedDevicesModalVisible" />
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/components/notifications/management/NotificationsManagementGeneralTab.vue]: Unfortunately this isn't possible: `pairedDevicesModalVisible` is set to false as soon as the user closes the modal. But the modal has a close animation afterwards which would be skipped with your approach.

File: frontend/components/notifications/management/NotificationsManagementGeneralTab.vue [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    pushToggle.value = g.is_push_notifications_enabled
+    muteTimestamp.value = g.do_not_disturb_timestamp > (Date.now() / 1000) ? g.do_not_disturb_timestamp : undefined
+  }
+}, { immediate: true })
+
+watch([emailToggle, pushToggle, muteTimestamp], ([enableEmail, enablePush, muteTs]) => {
+  if (!generalSettings.value) {
+    return
+  }
+  if (generalSettings.value?.is_email_notifications_enabled !== enableEmail || generalSettings.value?.is_push_notifications_enabled !== enablePush || generalSettings.value?.do_not_disturb_timestamp !== muteTs) {
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/components/notifications/management/NotificationsManagementGeneralTab.vue:76]: `watchEffect` triggers once instantly and we don't want that. I see your point, but for now I deem it good enough.

File: frontend/components/notifications/management/NotificationsManagementGeneralTab.vue [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+          {{ tOf($t, 'notifications.general.download_app', 1) }}
+        </BcLink>
+        {{ tOf($t, 'notifications.general.download_app', 2) }}
+      </div>
+    </div>
+    <div class="row">
+      <div>
+        {{ $t('notifications.general.send_test_email') }}
+      </div>
+      <Button class="p-button-icon-only" :disabled="testButtonsDisabled" @click="sendTestNotification('email')">
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/components/notifications/management/NotificationsManagementGeneralTab.vue:153]: I played around with this a bit but did not get a perfectly satisfying result. We can come back to this idea later imho.

File: frontend/components/notifications/management/NotificationsManagementGeneralTab.vue [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    </div>
+    <div class="row divider">
+      <div>
+        {{ $t('notifications.general.push_notifications') }}
+        <span v-if="pairedDevicesCount > 0">
+          ({{ pairedDevicesCount }})
+          <FontAwesomeIcon
+            class="link popout"
+            :icon="faArrowUpRightFromSquare"
+            @click="openPairdeDevicesModal"
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/components/notifications/management/NotificationsManagementGeneralTab.vue:136]: As discussed, I will create an internal ticket for this as we already have various locations that violate those rules.

File: frontend/components/notifications/management/NotificationsManagementPairedDevicesModal.vue [hunk 6]
```diff
@@ -0,0 +1,57 @@
+<script lang="ts" setup>
+import { useNotificationsManagementSettings } from '~/composables/notifications/useNotificationsManagementSettings'
+
+const { t: $t } = useI18n()
+
+const visible = defineModel<boolean>()
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/components/notifications/management/NotificationsManagementPairedDevicesModal.vue:6]: For this very instance I will not change the name as this is the same name used by PrimeVUE, see [HERE](https://primevue.org/dialog/#api.dialog.props.visible).

File: frontend/components/notifications/management/NotificationsManagementPairedDeviceModalContent.vue
Comment (marcel-bitfly) [inline frontend/components/notifications/management/NotificationsManagementPairedDeviceModalContent.vue:1]: this is a `nested dialog` hence if I hit `esc` the event has to be captured, e.g.: 

`@keydown.esc.stop.prevent="close"` (.prevent might be left out)

For the capturing to work, the focus has to be on a child element. The focus on a dialog element, in general should be on the `first interactive element`. In this case I would put it on the `close action` (done button).

File: frontend/components/notifications/management/NotificationsManagementPairedDeviceModalContent.vue [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+<template>
+  <div class="row-container">
+    <div class="device-row ">
+      <div class="device">
+        {{ $t('notifications.general.paired_devices.device') }}: {{ props.device.name || $t('notifications.general.paired_devices.unknown') }}
+      </div>
+      <Button data-secondary class="p-button-icon-only">
+        <FontAwesomeIcon :icon="faTrash" />
+      </Button>
```
Comment (marcel-bitfly) [inline frontend/components/notifications/management/NotificationsManagementPairedDeviceModalContent.vue:28]: tell me what this button does
1. best way: `button-text` that is only perceivable via screen reader, see (sr-only): https://tailwindcss.com/docs/screen-readers
2. aria-label='Permanently remove 'My device' from the list.' 
3. title-attribute for svg

**Comment by marcel-bitfly** [comment]
how should the dialog button behave, when there are no `paired devices` and `Email notifications` are turned off?

image

File: frontend/components/notifications/management/NotificationsManagementPairedDeviceModalContent.vue (hunk 7 above)
Comment (D13ce) [reply frontend/components/notifications/management/NotificationsManagementPairedDeviceModalContent.vue:28]: Will be done in a separate ticket, just like mentioned here: https://github.com/gobitfly/beaconchain/pull/594#discussion_r1696747034

File: frontend/components/notifications/management/NotificationsManagementGeneralTab.vue [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+      {
+        summary: $t('notifications.general.test_notification_error.toast_title'),
+        group: $t('notifications.general.test_notification_error.toast_group'),
+        detail: $t('notifications.general.test_notification_error.toast_message')
+      })
+  }
+  bounceTestButton(false)
+}
+
+const pairedDevicesCount = computed(() => pairedDevices.value?.length || 0)
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/components/notifications/management/NotificationsManagementGeneralTab.vue:58]: Will not change for now.

File: frontend/components/notifications/management/NotificationsManagementGeneralTab.vue [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+          <FontAwesomeIcon
+            class="link popout"
+            :icon="faArrowUpRightFromSquare"
+            @click="openPairdeDevicesModal"
+          />
+        </span>
+      </div>
+      <BcToggle v-if="pairedDevicesCount > 0" v-model="isPushToggleOn" />
+      <div v-else>
+        {{ tOf($t, 'notifications.general.download_app', 0) }}
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/components/notifications/management/NotificationsManagementGeneralTab.vue:142]: Will not change as this is not the only occurrence of this "pattern". Feel free to create a ticket to get all of those cleaned up.

File: frontend/components/notifications/management/NotificationsManagementPairedDeviceModalContent.vue
Comment (D13ce) [reply frontend/components/notifications/management/NotificationsManagementPairedDeviceModalContent.vue:1]: Will be handled in an internal ticket.

**Review by marcel-bitfly — APPROVED** [review]
have a good one

**Comment by D13ce** [comment]
> have a good one

Thank you :heart_hands:

**Pull request merged**